src/OrleansRuntime/GrainDirectory/LocalGrainDirectory.cs

```diff
@@ -469,175 +472,235 @@ public SiloAddress CalculateTargetSilo(GrainId grain, bool excludeThisSiloIfStop
+                        else
+                        {
+                            var newcached = new List<Tuple<SiloAddress, ActivationId>>(cached.Count + 1);
+                            newcached.AddRange(cached);
+                            newcached.Add(Tuple.Create(address.Silo, address.Activation));
+                            cached = newcached;
+                        }
+                        // update the cache so next local lookup will find this ActivationAddress in the cache and we will save full lookup.
+                        DirectoryCache.AddOrUpdate(address.Grain, cached, result.VersionTag);
                     }
-                    else
-                    {
-                        var newcached = new List<Tuple<SiloAddress, ActivationId>>(cached.Count + 1);
-                        newcached.AddRange(cached);
-                        newcached.Add(Tuple.Create(address.Silo, address.Activation));
-                        cached = newcached;
-                    }
-                    // update the cache so next local lookup will find this ActivationAddress in the cache and we will save full lookup.
-                    DirectoryCache.AddOrUpdate(address.Grain, cached, eTag);
                 }
-            }
-        }
 
-        public Task UnregisterAsync(ActivationAddress addr)
-        {
-            return UnregisterAsyncImpl(addr, true);
+                return result;
+            }
         }
 
         public Task UnregisterConditionallyAsync(ActivationAddress addr)
         {
             // This is a no-op if the lazy registration delay is zero or negative
             return Silo.CurrentSilo.OrleansConfig.Globals.DirectoryLazyDeregistrationDelay <= TimeSpan.Zero ? 
-                TaskDone.Done : UnregisterAsyncImpl(addr, false);
+                TaskDone.Done : UnregisterAsync(addr, false, 0);
         }
 
-        private Task UnregisterAsyncImpl(ActivationAddress addr, bool force)
+        public async Task UnregisterAsync(ActivationAddress address, bool force, int hopCount)
         {
-            unregistrationsIssued.Increment();
-            SiloAddress owner = CalculateTargetSilo(addr.Grain);
-            if (owner == null)
+            (hopCount > 0 ? UnregistrationsRemoteReceived : unregistrationsIssued).Increment();
+
+            if (hopCount == 0)
+                InvalidateCacheEntry(address);
+
+            // see if the owner is somewhere else (returns null if we are owner)
+            var forwardaddress = this.CheckIfShouldForward(address.Grain, hopCount, "UnregisterAsync");
+
+            // on all silos other than first, we insert a retry delay and recheck owner before forwarding
+            if (hopCount > 0 && forwardaddress != null)
             {
-                // We don't know about any other silos, and we're stopping, so throw
-                throw new InvalidOperationException("Grain directory is stopping");
+                await Task.Delay(RETRY_DELAY);
```

Comment:
Same answer. It was a common pattern to have two methods in LocalGrainDirectory and RemoteGrainDirectory that do almost the same thing, but the one in RemoteGrainDirectory inserts a delay before forwarding. These method pairs are now all deduplicated.

I realize that  this deduplication is really hard to understand in a diff, since you can't see where the code came from.